File: java/proverb/src/main/java/Proverb.java
class Proverb {
    String[] list;

    Proverb(String[] words) {
        list = words;
    }

    public String recite() {
        if (list.length == 0) {
            return "";
        }

        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < list.length; i++) {
            if (i <= 0 || i + 1 > list.length) {
                continue;
            }
                
            sb.append(String.format("For want of a %s the %s was lost.\n", list[i-1], list[i]));
        }
        sb.append(String.format("And all for the want of a %s.", list[0]));

        return sb.toString();
    }

}

File: csharp/proverb/Proverb.cs
using System;
using System.Collections.Generic;

public static class Proverb
{
    public static string[] Recite(string[] subjects)
    {
        if (subjects.Length == 0)
        {
            return subjects;
        }

        var ret = new List<string>();

        for (int i = 0; i < subjects.Length - 1; i++)
        {
            ret.Add($"For want of a {subjects[i]} the {subjects[i + 1]} was lost.");
        }

        ret.Add($"And all for the want of a {subjects[0]}.");

        return ret.ToArray();
    }
}